Promote "Metadata in Table Schema" recipe to the specs.

pierrecamilleri opened the pull request:
Context and related issue : 
fix [#899](https://github.com/frictionlessdata/specs/issues/899#issuecomment-2076509819)

# Details 

I tried to be as clear as possible in the constraints associated to the different metadata.

I chose to make a specific section for these properties, as they specifically refer to the schema and not to the data the schema describes, contrary to the other top schema properties. One may argue that the "$schema" property can also be seen as metadata, and could be moved into this section. What do you think ?

I see two points worth mentioning, as they differ from what was in the recipe : 

- The proposal adds more constraints on the `version` field and impose semantic versioning (with or without "v" prefix)
- The proposal is more restrictive in the `name` property, and disallows the underscore "_" character. This is aiming at a more URL-friendly name. The current implementation of frictionless allows this character, but as the property was to my knowledge not in the specs, I don't think this can be considered a breaking change.

I'm happy to listen to any requests for modifications.

pschumm commented:
I'm generally positive about this, and certainly can appreciate its value. One concern is that such expansions to the official specification don't make it less accessible, which I think this PR does in part by placing these additional properties in a separate section. There might be other things that can be done in this regard in the final documentation.

One question: Is there a place where we track the linkage between these metadata properties and other established standards such as Dublin Core? That would make me a bit more comfortable when adding new properties, and would be a great resource when advocating for Frictionless in various contexts.

peterdesmet commented:
I generate HTML documentation from Table Schemas in (using Jekyll) in several places ([here](https://camtrap-dp.tdwg.org/data/), [here](https://aloftdata.eu/vpts-csv/) and [here](https://esas-docs.ices.dk/tables/)), so I support more structured metadata.

- Personally, I only needed `title`, `name` and `description` as extra attributes.
- Before reviewing, I'd like to see the changes to [schema.yaml](https://github.com/frictionlessdata/datapackage/blob/main/profiles/source/dictionary/schema.yaml) and how it borrows from the common dictionary. @roll is this something you can add?
- `version` can be useful from a metadata perspective, but note that it can conflict with how we want to handle versioning when referring to a Table Schema. For example: a table schema at `https://example.org/1.0/my-table-schema.json` (implied to be version 1.0) could contain `version: 0.9`.
- @pschumm "linkage between these metadata properties and other established standards such as Dublin Core" is true for all properties in Data Package and are maybe better defined at the [common dictionary level](https://github.com/frictionlessdata/datapackage/blob/main/profiles/source/dictionary/common.yaml), so they apply to all Data Package elements that use them. In Camtrap DP, I have added those as `skos:exactMatch` (or alike) properties, see [here](https://github.com/tdwg/camtrap-dp/blob/188d285a3765f379f7c16ac2f542090cc1dfc86a/camtrap-dp-profile.json#L255). Maybe we should open an issue to do discuss this further?

pschumm commented:
> * @pschumm "linkage between these metadata properties and other established standards such as Dublin Core" is true for all properties in Data Package and are maybe better defined at the [common dictionary level](https://github.com/frictionlessdata/datapackage/blob/main/profiles/source/dictionary/common.yaml), so they apply to all Data Package elements that use them. In Camtrap DP, I have added those as `skos:exactMatch` (or alike) properties, see [here](https://github.com/tdwg/camtrap-dp/blob/188d285a3765f379f7c16ac2f542090cc1dfc86a/camtrap-dp-profile.json#L255). Maybe we should open an issue to do discuss this further?

Thanks @peterdesmet; I have just done so with #69.

roll commented on content/docs/specifications/table-schema.md:
```diff
@@ -293,6 +293,86 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+In addition, a Table Schema descriptor `MAY` contain these additional optional standard properties giving contextual information to the schema:
+
+#### `name`
```
I would just copy applicable parts of `resource.name` definition - https://datapackage.org/specifications/data-resource/#name-required (in the future we might move it to the Glossary)

pierrecamilleri replied:
Does it seem ok for you to *not* allow dots `.` and underscores `_` (contrary to the data-resource specification), or would you'd like to harmonize the two ? 

This suggestion was about "URL friendliness", but not sure it is really worth different behaviors.

roll replied:
@peterdesmet WDYT?

peterdesmet replied:
I would align the two (same rules for Table Schema and Data Resource `name`).

A Table Schema is going to be referenced to via URL in `$schema` and not `name`, so I don't think it has to be URL friendly.

pierrecamilleri replied:
Thanks for your comment. 
The `$schema` property will reference the Table Schema profile ; the idea behind making "name" URL-friendly is to have a systematic way to distribute the schemas, e.g. `https://my.catalog.com/{schema name}`, which is not quite the same. 

That said, I have no objection to either solution.

roll commented on content/docs/specifications/table-schema.md:
```diff
@@ -293,6 +293,86 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+If present, it `MUST` be a non-empty and lower-case string, containaing only alphanumeric characters along with the dash "-" character. It `MUST` not exceed 100 characters.
+
+#### `title`
```
I would sync with `package.title` - https://datapackage.org/specifications/data-package/#title (actually `package.title` might need to be extended as well)

I would drop 100 chars limit as it might be too restrictive

roll commented on content/docs/specifications/table-schema.md:
```diff
@@ -293,6 +293,86 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+If present, it `MUST` be a non-empty string, and not exceed 100 characters.
+
+#### `description`
```
Similarly to above I would sync with `package.description` - https://datapackage.org/specifications/data-package/#description

roll commented on content/docs/specifications/table-schema.md:
```diff
@@ -293,6 +293,86 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+If present, it `MUST` be a non-empty string.
+
+#### `homepage`
```
I think it would be nice to update `package.homepage` as well to make them in-sync

roll commented on content/docs/specifications/table-schema.md:
```diff
@@ -293,6 +293,86 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+If present, it `MUST` be a fully qualified URL as described in [RFC 1738](https://www.ietf.org/rfc/rfc1738.txt). 
+
+#### `path`
```
I'm not sure if we can standardize it at least under this name. It's kinda self-referencing. Maybe it can be something like `canonicalPath`?

pierrecamilleri replied:
I am not sure about "canonicalPath", as this has some meaning for POSIX paths 
(normalized absolute path). 

However I have no better proposition to make :s

roll commented on content/docs/specifications/table-schema.md:
```diff
@@ -293,6 +293,86 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+#### `path`
+
+The `homepage` property gives a direct path to the schema itself.
```
Typo `homepage` -> `path`

roll commented on content/docs/specifications/table-schema.md:
```diff
@@ -293,6 +293,86 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+If present, it `MUST` be a fully qualified URL as described in [RFC 1738](https://www.ietf.org/rfc/rfc1738.txt). The response body content `MUST` be exactly the schema.
+
+#### `sources`
```
I would just refer `package.sources` the same way we do in https://datapackage.org/specifications/data-resource/#sources

roll commented on content/docs/specifications/table-schema.md:
```diff
@@ -293,6 +293,86 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+If present, it `MUST` be a non-empty array of objects. Each object `MUST` have a "title" and a "path" property. The "title" properties are human-readable strings. The "path" property is the fully-qualified URL to the resource, it `MUST` be a fully qualified URL as described in [RFC 1738](https://www.ietf.org/rfc/rfc1738.txt).
+
+#### `created`
```
What if we just borrow a simplified DataCite's `dates` here for created/updated/etc - https://datacite-metadata-schema.readthedocs.io/en/4.5/properties/date/? Can be later used across the specs

If not, I would go with something like `dateCreated` and `dateUpdated`

pierrecamilleri replied:
DataCite's `dates` would enable to embed a whole changelog into the schema, why not, but I agree that a bit of simplification would be welcome (it is very "publisher" oriented). I would suggest to keep "Created", "Updated" and "Other".  

What do you think about : 

```markdown
### `dates`

The `dates` property contains an array of dates relevant to the Schema. 

If present, each date `MUST` be an object containing following properties : 

- the `date` property `MUST` be an [ISO 8601](https://www.iso.org/iso-8601-date-and-time-format.html) date (YYYY-MM-DD).
- the `dateType` property is a string that `MUST` take one of the following values : "Created", "Updated" or "Other".
- an optional `dateInformation` property, to store additional free-text information
```

roll replied:
My main concern with DataCite's `date` is very broad date-time definition

pierrecamilleri replied:
Yes, I don't think it's a good idea to accept all these formats. It would complicate implementations considerably. I proposed a single date format above, but I see that the "created" property of datapackage requires a datetime. So maybe ISO date or ISO datetime ? Or only the most encompassing datetime ?

roll replied:
@peterdesmet Can you please help here?

peterdesmet replied:
I would only add `created` and follow the Data Package format (datatime with timezone) and description.

roll commented on content/docs/specifications/table-schema.md:
```diff
@@ -293,6 +293,86 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+If present, it `MUST` be an [ISO 8601](https://www.iso.org/iso-8601-date-and-time-format.html) date (YYYY-MM-DD).
+
+#### `version`
```
We can just reference `package.version` (moving to the Glossary later)

roll commented on content/docs/specifications/table-schema.md:
```diff
@@ -293,6 +293,86 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+       
+Examples: "2.3.0" or "v2.3.0" or "2.3.0-beta"
+ 
+#### `contributors`
```
I would just refer `package.contributors` the same way we do in https://datapackage.org/specifications/data-resource/#sources

roll commented on content/docs/specifications/table-schema.md:
```diff
@@ -293,6 +293,86 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+If present, it `MUST` be a non-empty array of strings.
+ 
+#### `resources`
```
I think we can't use the property `resources` here as it's cognitively reserved by the Data Resource object. In my opinion, `examples` would perfectly fit here

roll reviewed (commented):
Great work @pierrecamilleri!

Sorry for the slow response, looks good in-general, I added some comments

roll commented:
@pierrecamilleri 
Also, thanks to Peter and Phil's comments, a few more thoughts:
- we really don't need an additional section for it as it is just normal metadata the same level as `fields` or `missingValues`
- can you please update `profiles/source/dictionary/schema.yaml` and run `node profiles/build.js` to build the user-facing JSON Schemas?

pierrecamilleri commented:
Thanks, I tried my best to take all your comments into account. 

Some remarks : 

- I noticed that you have hidden the recipes, I don't know if I should delete the metadata recipe or not in this PR. 
- I had difficult times to factorize the JSON Schema, so I preferred to reuse some "common.json" definitions rather than duplicating the data, which can lead to not very accurate descriptions ("descriptor", or "resource" in place of "schema"). I had to adapt specifically some descriptions mentioning a data package, with this vocabulary ; tell me if this is not OK. 
- In particular, I am not sure about reusing the "sources" from "common.json". The description seems a bit cryptic to me ("The raw sources for this resource.") in the context of a Schema.

pierrecamilleri commented:
Thanks, I tried my best to take all your comments into account. 

Some remarks : 

- I noticed that you have hidden the recipes, I don't know if I should delete the metadata recipe or not in this PR. 
- I had difficult times to factorize the JSON Schema, so I preferred to reuse some "common.json" definitions rather than duplicating the data, which can lead to not very accurate descriptions ("descriptor", or "resource" in place of "schema"). I had to adapt specifically some descriptions mentioning a data package, with this vocabulary ; tell me if this is not OK. 
- In particular, I am not sure about reusing the "sources" from "common.json". The description seems a bit cryptic to me ("The raw sources for this resource.") in the context of a Schema.

roll commented:
> I noticed that you have hidden the recipes, I don't know if I should delete the metadata recipe or not in this PR

No problem, I'll handle it later

roll commented on content/docs/specifications/table-schema.md:
```diff
@@ -265,6 +265,90 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+The `contributors` property stores people or oganizations that contributed to 
+the schema, as for [Data 
+Package](https://datapackage.org/specifications/data-package/#contributors).  
+If not specified the schema inherits from the data package.
```
In this and similiar remarks I think it would also make sense to mention something like "if distributed in a data package descriptor" because it's it's a standalone Table Schema there is not data package

roll commented on content/docs/specifications/table-schema.md:
```diff
@@ -265,6 +265,90 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+If present, it `MUST` be a non-empty array of objects. Each object `MUST` have 
+a "title" and a "path" property. The "title" properties are human-readable 
+strings. The "path" property is the URL to the resource, it `MUST` be a fully 
```
"URL to the resources" -- I think as I reader I might confused as it's not clear what kind of resource

pierrecamilleri replied:
This property is for instance to have example data directly presented to the user when developing validation tools with UIs (it is the case for "Validata", e.g [link](https://validata.fr/table-schema?schema_name=schema-datagouvfr.CEREMA%2Fschema-aire-livraison)) instead of uploading a file, the user can select "exemple files", 3rd tab). It would not be possible only with `homepage`.  

> This is not how the example property is used in Table Schema. 

If you mean the Field [`example` property](https://datapackage.org/specifications/table-schema/#example), it seems to me like a natural continuation to present a whole file as an example for the property at the schema's top level. 

I acknowledge however that having two `example` properties with different behaviors depending on its position in the schema might add some friction.

pierrecamilleri replied:
I am not sure what bothers you with the `examples` property. It appears to me good practice to provide example data, and it seems not very specific to our use case. 

Would a rename into `exampleResources` lift your reservations ?

peterdesmet replied:
`examples` is fine as a name (we don't use `exampleValues` for Table Schema fields either) and you are right, it is broad enough for wider use, I think the definition needs clarifications.

Would describe as:

> Example [Data Resources](../data-resource/) for this Table Schema (both valid and invalid).

Questions:

- Why `title` required?
- Is `path` referencing the JSON object of a stand-alone Data Resource? Or the path to a CSV file. I couldn't tell from the `exemple-antibes.csv` in your example. I guess it should be a Data Resource, but in that case, it _could_ also be provided as an object directly.
- Is it necessary to indicate if the referenced Data Resource is valid or not? Not that Table Schema fields examples are expected to be valid.

pierrecamilleri replied:
> Why title required?

This may allow to highlight some features about the provided data, for instance a schema accepting partial or full records may link two examples with evocative titles. 

> Is path referencing the JSON object of a stand-alone Data Resource? Or the path to a CSV file. I couldn't tell from the exemple-antibes.csv in your example. I guess it should be a Data Resource, but in that case, it could also be provided as an object directly.

We used it to link to the data file, but you are right it would make much more sense to have a full data resource and its flexibility ("title" and "path" properties are actually a subset of the properties of a data resource, but "name" is missing to be a valid one).  I change the documentation accordingly. 

> Is it necessary to indicate if the referenced Data Resource is valid or not? Not that Table Schema fields examples are expected to be valid.

In our use case, we tell this inside the title. The downside is that this way, we could not validate the data when it is expected to be valid.

In my opinion, we do not want to validate the data (which would often imply fetching remote data), and the case of an example with invalid data may be the most specific part to our use-case (we sometimes use it as a demo of what may be validated, or if we notice common pitfalls). I would suggest just removing the mention of invalid data in the documentation.

roll commented:
Great work @pierrecamilleri!

Let's wait for voting to complete; my only partial concern currently, the date fields and their possible sync with Data Package

peterdesmet commented on content/docs/specifications/table-schema.md:
```diff
@@ -265,6 +265,90 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+#### `lastModified`
+       
+The `lastModified` property contains the date on which the scheam was last modified.
```
```suggestion
The `lastModified` property contains the date on which the schema was last modified.
```

peterdesmet commented on content/docs/specifications/table-schema.md:
```diff
@@ -265,6 +265,90 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+If present, it `MUST` be an [ISO 8601](https://www.iso.org/iso-8601-date-and-time-format.html) date (YYYY-MM-DD).
+
+#### `lastModified`
```
This property is not one that exists in Data Package and I would not add it here. I prefer versions over modification dates. For a new version, the `created` property can be reset.

pierrecamilleri replied:
This is not how the `created` property is described right now : 

> Note: semantics may vary between publishers -- for some this is the datetime the data was created, for others the datetime the package was created.

I am not fond of having such an ambiguous (or overloaded) property, but we could certainly live with it !

pierrecamilleri replied:
(for what it's worth, [schema.data.gouv](https://schema.data.gouv.fr/etalab/schema-amenagements-cyclables/0.3.4/CHANGELOG.html) shows the version, alongside a changelog where there is usually the version release dates, I check with the team but I think we can remove it without overloading "created")

peterdesmet replied:
Yeah, it would be good to have a more stringent definition for `created` (to me it is squarely "the datetime the package was created" 😄 ), I have created a separate issue: https://github.com/frictionlessdata/specs/issues/944

I'm would suggest not to introduce `lastModified` to not muddy the waters. Good that you have a workaround. 😄 Note also that you can keep the property for schema.data.gouv, I would just not introduce it in the main spec.

peterdesmet commented on content/docs/specifications/table-schema.md:
```diff
@@ -265,6 +265,90 @@ If the value of the `foreignKey.reference.resource` property is an empty string
+
+If present, it `MUST` be a fully qualified URL as described in [RFC 1738](https://www.ietf.org/rfc/rfc1738.txt). The response body content `MUST` be exactly the schema.
+
+#### `sources`
```
Thanks! I now notice that the suggested definition for `table-schema$sources` is:

> List of documentation sources as for [Data Package](https://datapackage.org/specifications/data-package/#sources). If not specified the resource inherits from the data package.

This is quite a bit wider in scope than what Data Package defines:

> The raw sources for this data package.

Personally, I wouldn't want to expand the scope to any related documentation or background material, which seem to be the case with the examples you list. Such relations can be better expressed as with a custom property, such as

- References (cf https://camtrap-dp.tdwg.org/metadata/#references): array of strings
- relatedIdentifiers (cf https://camtrap-dp.tdwg.org/metadata/#relatedIdentifiers), borrowed from DataCite: array of objects which can clarify the relationship, see [example](https://github.com/tdwg/camtrap-dp/blob/b089447cad6945a7862843647e8663c375dc5c7f/example/datapackage.json#L259-L272).

pierrecamilleri replied:
I removed the `sources` property from the Table Schema specs.

roll commented:
I think it already looks good but I have one question regarding `examples` - "The “path” property is the URL to the example resource, it MUST be a fully qualified URL as described in [RFC 1738](https://www.ietf.org/rfc/rfc1738.txt)." - I would say it's not clear what exactly is expected there. Is it a data file or a Data Resource descriptor?

roll commented:
@khughitt 
@PietrH 
@pschumm 
@ezwelty 
@nichtich 
Can you please take a look?

nichtich commented:
Looks good as far as I can judge (via mobile, on vacation).

roll commented:
Hi @pierrecamilleri, can you please update the changelog as well:

> It's under v2 / Table Schema heading

```
##### `name` (new)

##### `title` (new)

##### `description` (new)

##### `homepage` (new)

##### `keywords` (new)

##### `examples` (new)

##### `created` (new)

##### `version` (new)

##### `contributors` (new)

An adjusted [Metadata in Table Schema](https://specs.frictionlessdata.io/patterns/#table-schema-metadata-properties) pattern has been promoted to the standard ([#63](https://github.com/frictionlessdata/datapackage/pull/63)).
```

pierrecamilleri commented:
> I think it already looks good but I have one question regarding `examples` - "The “path” property is the URL to the example resource, it MUST be a fully qualified URL as described in [RFC 1738](https://www.ietf.org/rfc/rfc1738.txt)." - I would say it's not clear what exactly is expected there. Is it a data file or a Data Resource descriptor?

We used to use only data files, but it would make sense that it be a Data Resource descriptor.

I noticed however a technical issue : making it an array of data resources descriptors creates a circular reference in the json schema, as a data resource may itself reference a "schema".

I tried to push some changes ([this commit](https://github.com/frictionlessdata/datapackage/pull/63/commits/cc10899d176132160da95ef285b4639b2d165445)) that references a Data Resource Descriptor *without* the `schema` property, I am however afraid that I may have broken things : the expansion of "allOf" does not seem supported by `node profiles/build.js`. (See [here](https://github.com/frictionlessdata/datapackage/pull/63/commits/cc10899d176132160da95ef285b4639b2d165445#diff-1aaf3862601b57a81ff0031b8ac535fd80dddcd6fc1db980f6a345a3a00d926cR3))

Do you have a clue on how I could perform this ?

roll commented:
@pierrecamilleri 
TBH For me I would say that example being just data files makes more sense as it's more data publisher centric (a few people if any have some data resources published)

peterdesmet commented:
@pierrecamilleri I have a commit with some updated wording and fixing some links. I will commit this in a separate branch for you to review.

peterdesmet commented:
Hmm, git is all out of wack it seems. The repository has moved and the access rights have changed. Let's wait until the v2 dust has settled. Still to do on this PR:

- [ ] Move table schema to `/standards/table-schema.md` (cf. main)
- [ ] Move changelog to `/overview/changelog.md` (cf. main)
- [ ] Update wording so it aligns with Data Package (my pending PR)
- [ ] Update some links
- [ ] Fix invalid json example
- [ ] Clarify if `examples` are borrowing `title` and `path` from Data Resource (I think they are, included in my PR) or if this is something completely separate.

peterdesmet commented:
Thanks for your patience @pierrecamilleri 😄

roll commented:
@@pierrecamilleri
Thanks!
Unfortunately, we must migrate this to the merged spec/datapackage site - https://github.com/frictionlessdata/datapackage. I can assist but it will be great to keep author credits

pierrecamilleri commented:
I rebased this PR on the datapackage repo (see link  right above). Only conflict I had was the moved changelog, hope nothing else has broken.  

I also moved the checklist in the pull request description.
